elk/training/train.py

```diff
@@ -11,25 +12,97 @@
+    """Worker function for training reporters in parallel."""
+
+    while not input_q.empty():
+        args.device = torch.device(f"cuda:{rank}")
+
+        i, *data = input_q.get()
+        out_q.put((i, *train_reporter(args, i, *data)))
+
+
+def train_reporter(
+    args: Namespace,
+    layer_index: int,
+    train_h: torch.Tensor,
+    val_h: torch.Tensor,
+    train_labels: torch.Tensor,
+    val_labels: torch.Tensor,
+):
+    """Train a single reporter on a single layer."""
+
+    # Note: currently we're just upcasting to float32 so we don't have to deal with
+    # grad scaling (which isn't supported for LBFGS), while the hidden states are
+    # saved in float16 to save disk space. In the future we could try to use mixed
+    # precision training in at least some cases.
+    x0, x1 = train_h.to(args.device).float().chunk(2, dim=-1)
+    val_x0, val_x1 = val_h.to(args.device).float().chunk(2, dim=-1)
+
+    train_labels_aug = torch.cat([train_labels, 1 - train_labels])
+    val_labels_aug = torch.cat([val_labels, 1 - val_labels])
+
+    reporter = Reporter(
+        in_features=x0.shape[-1],
+        device=args.device,
+        init=args.init,
+        loss=args.loss,
+        supervised_weight=args.supervised_weight,
+    )
+    if args.label_frac:
+        num_labels = round(args.label_frac * len(train_labels))
+        labels = train_labels[:num_labels].to(args.device)
+    else:
+        labels = None
+
+    train_loss = reporter.fit(
+        contrast_pair=(x0, x1),
+        labels=labels,
+        num_tries=args.num_tries,
+        optimizer=args.optimizer,
+        weight_decay=args.weight_decay,
+    )
+    val_result = reporter.score(
+        (val_x0, val_x1),
+        val_labels.to(args.device),
+    )
+
+    output_dir = elk_cache_dir() / args.name
+    lr_dir = output_dir / "lr_models"
+    reporter_dir = output_dir / "reporters"
+
+    lr_dir.mkdir(parents=True, exist_ok=True)
+    reporter_dir.mkdir(parents=True, exist_ok=True)
+    stats = [train_loss, *val_result]
+
+    if not args.skip_baseline and not args.num_devices > 1:
+        # TODO: Once we implement cross-validation for CCS, we should benchmark
+        # against LogisticRegressionCV here.
+        lr_model = LogisticRegression(max_iter=10_000)
+        lr_model.fit(torch.cat([x0, x1]).cpu(), train_labels_aug)
+
+        lr_preds = lr_model.predict_proba(torch.cat([val_x0, val_x1]).cpu())[:, 1]
+        lr_acc = accuracy_score(val_labels_aug, lr_preds > 0.5)
+        lr_auroc = roc_auc_score(val_labels_aug, lr_preds)
 
-    Contains the main training loop. Functionality varies based on
-    the arguments passed in.
-    Consult parser.py for more details on the arguments.
+        stats += [lr_auroc, lr_acc]
+        with open(lr_dir / f"layer_{layer_index}.pkl", "wb") as file:
+            pickle.dump(lr_model, file)
 
-    Args:
-        args: The arguments for training.
-    """
+    with open(reporter_dir / f"layer_{layer_index}.pt", "wb") as file:
```

Comment:
this saving convention in particular is what's confusing--"layer_{layer_index}.pkl" doesn't necessarily hold the data for layer `layer_index`

Reply:
Added a warning in `elk extract` when you try to save only a subset of layers as a bandaid solution